### PR TITLE
RAC-123: [Feature branch] Attribute & properties values translators

### PR DESCRIPTION
**Pull request by ValentinMumble**
**Description (for Contributor and Core Developer)**

Mini feature branch for attribute & properties values.
All PRs merged here have been GTMed by either @SamirBoulil or @ValentinMumble, we are waiting for the rest of the team to review them next week.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | OK
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Query/SqlGetProductLabels.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+{
+    /** @var Connection */
+    private $connection;
+
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
+
+    public function byCodesAndLocaleAndScope(array $identifiers, string $locale, string $channel): array
```
Comment (juliensnz) [inline src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Query/SqlGetProductLabels.php:18]: byCodes but it takes identifiers?

**Review by juliensnz — APPROVED** [review]

**Review by SamirBoulil — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            $translations = $this->getProductModelLabels->byCodesAndLocaleAndScope($codes, $locale, $channel);
+        } elseif ($this->isProductAssociation($columnName)) {
+            $translations = $this->getProductLabels->byCodesAndLocaleAndScope($codes, $locale, $channel);
+        } elseif ($this->isGroupAssociation($columnName)) {
+            $translations = $this->getGroupTranslations->byGroupCodesAndLocale($codes, $locale);
+        } else {
+            throw new \LogicException(sprintf('Unsupported column to translate associations "%s"', $columnName));
+        }
+
+        return $translations;
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php:75]:
```suggestion
        if ($this->isProductModelAssocation($columnName)) {
            return $this->getProductModelLabels->byCodesAndLocaleAndScope($codes, $locale, $channel);
        } 
        
        if ($this->isProductAssociation($columnName)) {
            return $this->getProductLabels->byCodesAndLocaleAndScope($codes, $locale, $channel);
        } 
        
        if ($this->isGroupAssociation($columnName)) {
            return $this->getGroupTranslations->byGroupCodesAndLocale($codes, $locale);
        }

        throw new \LogicException(sprintf('Unsupported column to translate associations "%s"', $columnName));
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    {
+        $associationsColumns = $this->associationColumnsResolver->resolveAssociationColumns();
+        $quantifiedAssociationsColumns = $this->associationColumnsResolver->resolveQuantifiedAssociationColumns();
+
+        return in_array($columnName, array_merge($associationsColumns, $quantifiedAssociationsColumns));
+    }
+
+    public function translate(string $columnName, array $values, string $locale, string $channel): array
+    {
+        if ($this->isColumnWithQuantity($columnName)) {
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php:52]: Maybe it's more relevant to call `resolveQuantifiedIdentifierAssociationColumns` instead  `resolveQuantifiedAssociationColumns` in `support` method 

```
    public function supports(string $columnName): bool
    {
        $associationsColumns = $this->associationColumnsResolver->resolveAssociationColumns();
        $quantifiedAssociationsColumns = $this->associationColumnsResolver->resolveQuantifiedIdentifierAssociationColumns();

        return in_array($columnName, array_merge($associationsColumns, $quantifiedAssociationsColumns));
    }
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php [hunk 4]
```diff
@@ -0,0 +1,147 @@
+<?php
+
+namespace Akeneo\Pim\Enrichment\Component\Product\Connector\FlatTranslator\PropertyValue;
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php:3]: It seems that this class does not extend FlatPropertyValueTranslatorInterface, maybe this class should be directly in FlatTranslator directory

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        $allCategoryCodes = [];
+        foreach ($values as $value) {
+            $categoryCodes = explode(',', $value);
+            $categoryCodesWithoutQuantities = array_map(
+                function (string $categoryCodeWithQuantities) {
+                    return preg_replace('/\|.*$/', '', $categoryCodeWithQuantities);
+                },
+                $categoryCodes
+            );
+            $allCategoryCodes = array_merge($allCategoryCodes, $categoryCodesWithoutQuantities);
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php:89]: I didn't understand this logic => data in value are just identifier separate by comma, it does not contain quantities with separate with pipe

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+            $codes = explode(',', $value);
+            $labelized = [];
+
+            foreach ($codes as $code) {
+                preg_match('/^(?<code>.*)\|(?<quantity>.*)$/', $code, $matches);
+                $quantity = $matches['quantity'] ?? null;
+                if (empty($quantity)) {
+                    $translation = $translations[$code] ??
+                        sprintf(FlatTranslatorInterface::FALLBACK_PATTERN, $code);
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php:112]: Same than : https://github.com/akeneo/pim-community-dev/pull/12622/files#diff-f5bed6c9de27935ed4de0b11bb78914fR89

File: tests/back/Pim/Enrichment/Specification/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslatorSpec.php [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        );
+    }
+
+    function it_translates_product_codes_with_quantities_to_labels(GetProductLabelsInterface $getProductLabels)
+    {
+        $getProductLabels->byCodesAndLocaleAndScope(['tshirt', 'group-with-no-label'], 'fr_FR', 'ecommerce')
+            ->willReturn(['tshirt' => 'Groupe des T-shirts']);
+        $this->translate(
+            'X_SELL-products-quantified',
+            ['tshirt|2,group-with-no-label|1', 'group-with-no-label|1'],
```
Comment (StevenVAIDIE) [inline tests/back/Pim/Enrichment/Specification/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslatorSpec.php:76]: I never seen this syntax

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Query/SqlGetProductLabels.php (hunk 1 above)
Comment (SamirBoulil) [reply src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Query/SqlGetProductLabels.php:18]: done

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Query/SqlGetProductLabels.php [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        $query = <<<SQL
+SELECT
+    p.identifier,
+    a.code as label_code, 
+    a.is_localizable as label_is_localizable, 
+    a.is_scopable AS label_is_scopable,
+    JSON_MERGE_PATCH(COALESCE(p.raw_values, '{}'), COALESCE(pm1.raw_values, '{}'), COALESCE(pm.raw_values, '{}')) as raw_values
```
[the remark this answers is not in the document]
Comment (SamirBoulil) [reply src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Query/SqlGetProductLabels.php:26]: Done

File: tests/back/Pim/Enrichment/Specification/Component/Product/Connector/FlatTranslator/AttributeValuesFlatTranslatorSpec.php [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+            'scope_code'  => 'ecommerce'
+        ]);
+        $name->getCode()->willReturn('name');
+        $name->getType()->willReturn('pim_catalog_text');
+        $name->getProperties()->willReturn(['my_property' => 'not null value']);
+
+        $attributeValueRegistry->getTranslator('pim_catalog_text', 'name-en_US-ecommerce')->willReturn($translator);
+        $translator->translate('name', ['my_property' => 'not null value'], ['value'], 'fr_FR')->willReturn(['valeur une', 'valeur deux']);
+
+        $this->translate('name-en_US-ecommerce', ['value'], 'fr_FR')->shouldReturn(['valeur une', 'valeur deux']);
```
[the remark this answers is not in the document]
Comment (SamirBoulil) [reply tests/back/Pim/Enrichment/Specification/Component/Product/Connector/FlatTranslator/AttributeValuesFlatTranslatorSpec.php:74]: done

File: tests/back/Pim/Enrichment/Specification/Component/Product/Connector/FlatTranslator/PropertyValue/CategoryTranslatorSpec.php [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        $this->supports('categories')->shouldReturn(true);
+        $this->supports('other')->shouldReturn(false);
+    }
+
+    function it_translates_category_property_values(
+        GetCategoryTranslations $getCategoryTranslations
+    ) {
+        $getCategoryTranslations->byCategoryCodesAndLocale(['men_shoe', 'tshirt_women', 'not_translated_category'], 'fr_FR')->willReturn(['men_shoe' => 'Chaussures pour hommes', 'tshirt_women' => 'Tshirt pour femmes']);
+
+        $this->translate(['men_shoe,tshirt_women,not_translated_category'], 'fr_FR', 'ecommerce')->shouldReturn(['Chaussures pour hommes,Tshirt pour femmes,[not_translated_category]']);
```
[the remark this answers is not in the document]
Comment (SamirBoulil) [reply tests/back/Pim/Enrichment/Specification/Component/Product/Connector/FlatTranslator/PropertyValue/CategoryTranslatorSpec.php:27]: done

File: tests/back/Pim/Enrichment/Specification/Component/Product/Connector/FlatTranslator/AttributeValue/MultiSelectFlatTranslatorSpec.php [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+            ->willReturn(
+                [
+                    $this->optionKey('color', 'red')    => [$locale => 'rouge'],
+                    $this->optionKey('color', 'yellow') => [$locale => 'jaune'],
+                    $this->optionKey('color', 'purple') => [$locale => 'purple']
+                ]
+            );
+
+        $this->translate('color', [], ['red,yellow', 'purple'], $locale)
+            ->shouldReturn(['rouge,jaune', 'purple']);
```
[the remark this answers is not in the document]
Comment (SamirBoulil) [reply tests/back/Pim/Enrichment/Specification/Component/Product/Connector/FlatTranslator/AttributeValue/MultiSelectFlatTranslatorSpec.php:47]: fixed

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/AttributeValue/MultiSelectFlatTranslator.php [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+use Akeneo\Pim\Structure\Component\AttributeTypes;
+use Akeneo\Pim\Structure\Component\Query\PublicApi\AttributeOption\GetExistingAttributeOptionsWithValues;
+
+/**
+ * @author    Samir Boulil <samir.boulil@akeneo.com>
+ * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+
+class MultiSelectFlatTranslator implements FlatAttributeValueTranslatorInterface
```
[the remark this answers is not in the document]
Comment (SamirBoulil) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/AttributeValue/MultiSelectFlatTranslator.php:17]: done

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/AttributeValue/SimpleSelectFlatTranslator.php [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+use Akeneo\Pim\Enrichment\Component\Product\Connector\FlatTranslator\FlatTranslatorInterface;
+use Akeneo\Pim\Structure\Component\AttributeTypes;
+use Akeneo\Pim\Structure\Component\Query\PublicApi\AttributeOption\GetExistingAttributeOptionsWithValues;
+
+/**
+ * @author    Samir Boulil <samir.boulil@akeneo.com>
+ * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class SimpleSelectFlatTranslator implements FlatAttributeValueTranslatorInterface
```
[the remark this answers is not in the document]
Comment (SamirBoulil) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/AttributeValue/SimpleSelectFlatTranslator.php:16]: done

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php (hunk 2 above)
Comment (SamirBoulil) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php]: Done

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php (hunk 5 above)
Comment (SamirBoulil) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php:91]: I remove it

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php (hunk 4 above)
Comment (SamirBoulil) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php:5]: I'll move it

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php (hunk 3 above)
Comment (SamirBoulil) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslator.php]: Removed the function. not needed

File: tests/back/Pim/Enrichment/Specification/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslatorSpec.php (hunk 7 above)
Comment (SamirBoulil) [reply tests/back/Pim/Enrichment/Specification/Component/Product/Connector/FlatTranslator/PropertyValue/AssociationTranslatorSpec.php]: removed

File: tests/back/Pim/Enrichment/Specification/Component/Product/Connector/FlatTranslator/AttributeValue/BooleanTranslatorSpec.php [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+    function it_translates_boolean_attribute_values(
+        LabelTranslatorInterface $labelTranslator
+    ) {
+        $labelTranslator->translate('pim_common.yes', 'fr_FR', '[yes]')->willReturn('Oui');
+        $labelTranslator->translate('pim_common.yes', 'UNKNOWN', '[yes]')->willReturn('[yes]');
+        $labelTranslator->translate('pim_common.no', 'fr_FR', '[no]')->willReturn('Non');
+        $labelTranslator->translate('pim_common.no', 'UNKNOWN', '[no]')->willReturn('[no]');
+
+        $this->translate('with', [], ['1', '0', '1'], 'fr_FR')->shouldReturn(['Oui', 'Non', 'Oui']);
```
[the remark this answers is not in the document]
Comment (SamirBoulil) [reply tests/back/Pim/Enrichment/Specification/Component/Product/Connector/FlatTranslator/AttributeValue/BooleanTranslatorSpec.php:31]: done

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/AssociationTranslator.php [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+        $this->associationColumnsResolver = $associationColumnsResolver;
+        $this->getProductModelLabels = $getProductModelLabels;
+        $this->getProductLabels = $getProductLabels;
+        $this->getGroupTranslations = $getGroupTranslations;
+    }
+
+    public function supports(string $columnName): bool
+    {
+        $associationsColumns = $this->associationColumnsResolver->resolveAssociationColumns();
+        $quantifiedAssociationsColumns = $this->associationColumnsResolver->resolveQuantifiedAssociationColumns();
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/AssociationTranslator.php:46]:
```suggestion
        $quantifiedAssociationsColumns = $this->associationColumnsResolver->resolveQuantifiedIdentifierAssociationColumns();
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/AttributeValue/MetricTranslator.php [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    public function __construct(SqlGetUnitTranslations $getUnitTranslations)
+    {
+        $this->getUnitTranslations = $getUnitTranslations;
+    }
+
+    public function supports(string $attributeType, string $columnName): bool
+    {
+        $endsWithUnit = 0 === substr_compare($columnName, self::UNIT_SUFFIX, -strlen(self::UNIT_SUFFIX));
+
+        return $attributeType === AttributeTypes::METRIC && $endsWithUnit;
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/FlatTranslator/AttributeValue/MetricTranslator.php:27]:
```suggestion
        return $attributeType === AttributeTypes::METRIC && str_ends_with($columnName, self::UNIT_SUFFIX);
```

File: src/Akeneo/Pim/Enrichment/Bundle/Resources/config/connector/flat_translators/property_value_translators.yml [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+      - '@akeneo.pim.enrichment.product.query.get_product_model_labels'
+    tags: [{name: pim_enrich.connector.flat_translators.property_value}]
+
+  pim_enrich.connector.flat_translators.property_value_translator.associations:
+    class: Akeneo\Pim\Enrichment\Component\Product\Connector\FlatTranslator\AssociationTranslator
+    arguments:
+      - '@pim_connector.array_converter.flat_to_standard.product.association_columns_resolver'
+      - '@akeneo.pim.enrichment.product.query.get_product_model_labels'
+      - '@akeneo.pim.enrichment.product.query.get_product_labels'
+      - '@akeneo.pim.structure.query.get_group_translations'
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Bundle/Resources/config/connector/flat_translators/property_value_translators.yml:44]: Maybe this should be in `src/Akeneo/Pim/Enrichment/Bundle/Resources/config/connector/flat_translators.yml`

**Review by StevenVAIDIE — APPROVED** [review]

**Pull request merged**